Source/NSubstitute/Core/CallCollection.cs

```diff
@@ -0,0 +1,56 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace NSubstitute.Core
+{
+    public class CallCollection : ICallCollection, IReceivedCalls
+    {
+        ConcurrentQueue<CallWrapper> _callWrappers = new ConcurrentQueue<CallWrapper>();
+
+        public void Add(ICall call)
+        {
+            if (call == null) throw new ArgumentNullException(nameof(call));
+            _callWrappers.Enqueue(new CallWrapper(call));
+        }
+
+        public void Delete(ICall call)
+        {
+            if (call == null) throw new ArgumentNullException(nameof(call));
+
+            var callWrapper = _callWrappers.FirstOrDefault(w => w.Call.Equals(call));
+            if (callWrapper == null) throw new InvalidOperationException("Collection doesn't contain the call.");
+
+            callWrapper.IsDeleted = true;
```

Comment:
@dtchepak the only principle thing that was changed is this: we don't delete the "call" objects as we did in `CallStack`. We just mark it as deleted. In case of huge amount of specifications it could case OOM. I think it's better to change that to the previous behavior.